# Linear Algebra #1
> Vector norm and matrix multiplication

- toc: true
- branch: true
- badges: true
- comments: true
- author: Gui Osorio

## Vector Norm/Magnitude

Vector norm refers to calculating a vector's length. In practice, higher coefficients lead to less accurate models, therefore vector norm calculation is useful in ML, as it can be used in techniques to calibrate ML models -> minimizing the loss and preventing overfitting and underfitting. As the distance between two vectors is a single vector, claculating the resulting vector's norm can also be used as a loss function of a ML model.

Common methods to calculate a vector's norm are:
- L1 norm
- L2 norm (root mean squared error)
- Max norm

### L1 norm

Refers to the sum of absolute values in a vector. Summarizes the distance from the vector to its origin in space.

#### Single vector

In [32]:
#collapse-hide
import numpy as np
import torch
import torch.nn.functional as F

In [46]:
# Two methods to calculate L1 norm for a single vector as a NumPy array
# Define a vector
v = np.array([1,2,3])

# Calculate L1 from scrach
def calc_l1norm(vec): return float(abs(vec.sum()))
print('From scratch: ', calc_l1norm(v))

# Calculate L1 norm using NumPy
print('Using NumPy.linalg: ', np.linalg.norm(v, 1))

From scratch:  6.0
Using NumPy.linalg:  6.0


In the case above, the L1 norm can be calculated with the simple addition: 1 + 2 + 3 = 6

#### Two vectors

In [50]:
# Two me#### Example with a single vectorthods to calculate L1 norm for two vectors as PyTorch tensors
# Define 2 tensors
t1 = torch.tensor([1.,2.,3.])
t2 = torch.tensor([4.,5.,6.])

# Calculate L1 norm from scratch
def calc_l1norm2(tensor1, tensor2): return (tensor1-tensor2).abs().mean()
print('From scratch: ', calc_l1norm2(t1,t2))

# Calculate L1 norm using PyTorch
print('Using PyTorch: ', F.l1_loss(t1, t2))

From scratch:  tensor(3.)
Using PyTorch:  tensor(3.)


1. Subtract tensors -> t1 - t2 = [-3.,-3.,-3.]
2. Convert to absolute values -> [3.,3.,3.].
3. The L1 norm of this vector can then be calculated as taking the mean of these values -> (3+3+3)/3 = 3

### L2 norm

Refers to the root mean squared error of the values in a vector.

#### Single vector

In [62]:
# Calculating L2 norm in a single vector as a NumPy array
# Define a vector
v = np.array([1.,2.,3.])

# Calculate L2 from scratch
def calc_l2norm(vec): return np.sqrt(((vec)**2).sum())
print('From scratch: ', calc_l2norm(v))

# Calculate L2 norm using NumPy
print('Using NumPy.linalg: ', np.linalg.norm(v, 2))

From scratch:  3.7416573867739413
Using NumPy.linalg:  3.7416573867739413


1. The vector is transformed into its squared version [1**2, 2**2, 3**3] -> [1,4,9].
2. The squared vector is summed up 1 + 4 + 9 -> 14
3. The L2 is equal to the square root of the scalar sqrt(14) -> 3.742

#### Two vectors

In [61]:
# Calculating L2 norm between two vectors as PyTorch tensors
# Define 2 tensors
t1 = torch.tensor([1.,2.,3.])
t2 = torch.tensor([4.,5.,6.])

# Calculate L2 norm from scratch
def calc_l2norm2(tensor1, tensor2): return ((tensor1-tensor2)**2).mean().sqrt()
print('From scratch: ', calc_l2norm2(t1, t2))

# Calculate L2 norm using PyTorch
print('Using PyTorch: ', F.mse_loss(t1, t2).sqrt())

From scratch:  tensor(3.)
Using PyTorch:  tensor(3.)


1. Subtract tensors -> [1-4,2-5,3-6] -> [-3,-3,-3]
2. Square the subtracted tensor -> [-3**2,-3**2,-3**2] -> [9,9,9]
3. Calculate the mean as a scalar -> (9+9+9)/3 = 9
4. Calculate the square root of the scalar -> sqrt(9) = 3

### Max norm

Returns the maximum absolute value of the vector.

#### Single vector

In [68]:
# Calculate max norm for a single vector as a NumPy array
# Define vector
v = np.array([1,2,3])

# Calculate max norm from scratch
def calc_maxnorm(vec): return max(abs(vec))
print('From scratch: ', calc_maxnorm(v))

# Calculate max norm using NumPy.linalg
print('Using NumPy: ', np.linalg.norm(v, np.inf))

From scratch:  3
Using NumPy:  3.0


## Matrix Multiplication

Matrix multiplication is not so straight-forward as you may think. Below, I will explain the dot product between matrixes. The conventional multiplication you may be thinking of, when referred to vectors, is called the Hadamard multiplication.

### Matrix-Matrix

Let's say we have two matrixes, M1 and M2. Their dimensions (rows, columns) are (m,n) for M1 and (n,k) for M2.
For them to be multiplied, a simple requirement needs to be satisfied: n must be equal, that is, the number of columns in M1 needs to be equal to the number of rows in M2.
The resulting matrix will be of dimensions (m,k), that is, the number of rows of M1, and the number of columns of M2.

Let's represent this by multiplying a 2x3 matrix with a 3x2 matrix.

In [72]:
# Matrix-matrix multiplication
M1 = np.array([[1,2,3], [4,5,6]]) # 2x3 matrix
M2 = np.array([[1,2], [3,4], [5,6]]) # 3x2 matrix
print(M1)
print('------------')
print(M2)
M1 @ M2

[[1 2 3]
 [4 5 6]]
------------
[[1 2]
 [3 4]
 [5 6]]


array([[22, 28],
       [49, 64]])

As noted above, the resulting matrix has dimensions (2,2).

- Position [0,0] = M1[0,0] * M2[0,0] + M1[0,1] * M2[1,0] + M1[0,2] * M2[2,0] = 1 * 1 + 2 * 3 + 3 * 5 = 22
- Position [0,1] = M1[0,0] * M2[0,1] + M1[0,1] * M2[1,1] + M1[0,2] * M2[2,1] = 1 * 2 + 2 * 4 + 3 * 6 = 28
- Position [1,0] = M1[1,0] * M2[0,0] + M1[1,1] * M2[1,0] + M1[1,2] * M2[2,0] = 4 * 1 + 5 * 3 + 6 * 5 = 49
- Position [1,1] = M1[1,0] * M2[0,1] + M1[1,1] * M2[1,1] + M1[1,2] * M2[2,1] = 4 * 2 + 5 * 4 + 6 * 6 = 64
- Resulting in the 2x2 matrix [ [22,28], [49,64] ]

### Matrix-Vector

For a matrix-vector multiplication to work, the number of values in the vector needs to equal n, the number of columns in the matrix. Multiplying a matrix of dimensions (m, n) with a vector of k values results in a vector with m values (number of rows in matrix). 

In [76]:
v = np.array([1,2,3])
print(v)
print(M1)
M1 @ v

[1 2 3]
[[1 2 3]
 [4 5 6]]


array([14, 32])

- Position [0,0] = M1[0,0] * v[0] + M1[0,1] * v[1] + M1[0,2] * v[2] = 1 * 1 + 2 * 2 + 3 * 3 = 14
- Position [0,1] = M1[1,0] * v[0] + M1[1,1] * v[1] + M1[1,2] * v[2] = 4 * 1 + 5 * 2 + 6 * 3 = 32
- Resulting in a vector with 2 values [14, 32]